### PR TITLE
Fix logging to loggers with multiple eval dataloaders

### LINKED ISSUE #11907: Validation on epoch metrics not logging for multiple data loaders on 1.60dev
## 🐛 Bug

Used 2 dataloaders in validation and specified to log on epoch, but only seeing `val_loss_step` and no `val_loss_epoch` in Tensorboard.

```
def validation_step(self, batch, batch_idx, dataloader_idx):
        self._common_step(batch, batch_idx, f"val{dataloader_idx}")
        
 def _common_step(self, batch, batch_idx, stage: str):
        x = self._prepare_batch(batch)
        loss = F.mse_loss(x, self(x))
        self.log(f"{stage}_loss", loss, on_step=True, on_epoch=True)
        return loss       
```

Screen Shot 2022-02-13 at 9 28 56 PM

### To Reproduce

```
from typing import Optional, Tuple

import torch
import torch.nn.functional as F
from torch import nn
from torch.utils.data import DataLoader, random_split

import pytorch_lightning as pl
from pytorch_lightning.utilities.cli import LightningCLI
from pytorch_lightning.utilities.rank_zero import rank_zero_only
from tests.helpers.boring_model import BoringModel, RandomDataset

class LitAutoEncoder(pl.LightningModule):
    """
    >>> LitAutoEncoder()  # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
    LitAutoEncoder(
      (encoder): ...
      (decoder): ...
    )
    """

    def __init__(self, hidden_dim: int = 64):
        super().__init__()
        self.encoder = nn.Sequential(nn.Linear(100, hidden_dim), nn.ReLU(), nn.Linear(hidden_dim, 3))
        self.decoder = nn.Sequential(nn.Linear(3, hidden_dim), nn.ReLU(), nn.Linear(hidden_dim, 100))

    def forward(self, x):
        z = self.encoder(x)
        x_hat = self.decoder(z)
        return x_hat

    def training_step(self, batch, batch_idx):
        return self._common_step(batch, batch_idx, "train")

    def validation_step(self, batch, batch_idx, dataloader_idx):
        self._common_step(batch, batch_idx, f"val{dataloader_idx}")

    def test_step(self, batch, batch_idx):
        self._common_step(batch, batch_idx, "test")

    def predict_step(self, batch, batch_idx, dataloader_idx=None):
        x = self._prepare_batch(batch)
        return self(x)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

    def _prepare_batch(self, batch):
        # (batch_size, hidden_dim)
        return batch

    def _common_step(self, batch, batch_idx, stage: str):
        x = self._prepare_batch(batch)
        loss = F.mse_loss(x, self(x))
        self.log(f"{stage}_loss", loss, on_step=True, on_epoch=True)
        return loss


class MyDataModule(pl.LightningDataModule):
    def __init__(self, batch_size: int = 32):
        super().__init__()
        self.random_train = RandomDataset(100, 32)
        self.random_val = RandomDataset(100, 32)
        self.random_test = RandomDataset(100, 32)
        self.batch_size = batch_size

    def train_dataloader(self):
        return DataLoader(self.random_train, batch_size=self.batch_size)

    def val_dataloader(self):
        # use 2 data loaders
        l1 = DataLoader(self.random_val, batch_size=self.batch_size)
        l2 = DataLoader(self.random_val, batch_size=self.batch_size)
        return [l1, l2]

    def test_dataloader(self):
        return DataLoader(self.random_test, batch_size=self.batch_size)

    def predict_dataloader(self):
        return DataLoader(self.random_test, batch_size=self.batch_size)


def cli_main():
    cli = LightningCLI(
        LitAutoEncoder,
        MyDataModule,
        seed_everything_default=1234,
        save_config_overwrite=True,
        run=False,  # used to de-activate automatic fitting.
        trainer_defaults={'max_epochs': 3}
    )
    cli.trainer.fit(cli.model, datamodule=cli.datamodule)


if __name__ == "__main__":
    cli_main()
```

### Expected behavior

Log `val_loss_epoch` to Tensorboard

### Environment

* CUDA:
        - GPU:
        - available:         False
        - version:           10.2
* Packages:
        - numpy:             1.21.4
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.2+cu102
        - pytorch-lightning: 1.6.0dev
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                -
        - processor:         x86_64
        - python:            3.8.11
        - version:           #1 SMP Fri Jan 21 19:56:44 UTC 2022

### Additional context

N/A

cc @tchaton @rohitgr7 @akihironitta @carmocca @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #11907

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### CODE REVIEW

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 1]
```diff
@@ -765,13 +766,27 @@ def test_epoch_end(self, outputs) -> None:
...
         def test_dataloader(self):
-            return [super().val_dataloader(), super().val_dataloader()]
+            return [super().test_dataloader(), super().test_dataloader()]
 
     model = CustomBoringModel()
-    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_test_batches=1,
+    )
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:775]: nit:

```suggestion
    trainer = Trainer(default_root_dir=tmpdir, limit_test_batches=1)
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 2]
```diff
@@ -181,11 +181,13 @@ def on_run_end(self) -> List[_OUT_DICT]:
...
         # include any logged outputs on epoch_end
         epoch_end_logged_outputs = self.trainer._logger_connector.update_eval_epoch_metrics()
+        logger_metrics = dict(ChainMap(*logged_outputs))
+        logger_metrics.update(epoch_end_logged_outputs)
         for dl_outputs in logged_outputs:
             dl_outputs.update(epoch_end_logged_outputs)
 
         # log metrics
-        self.trainer._logger_connector.log_eval_end_metrics()
+        self.trainer._logger_connector.log_eval_end_metrics(logger_metrics)
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:190]: Wouldn't this work?

```suggestion
        all_logged_outputs = dict(ChainMap(*logged_outputs))  # list[dict] -> dict  
        self.trainer._logger_connector.log_eval_end_metrics(all_logged_outputs)
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 3]
```diff
@@ -181,11 +181,13 @@ def on_run_end(self) -> List[_OUT_DICT]:
         logged_outputs, self._logged_outputs = self._logged_outputs, []  # free memory
         # include any logged outputs on epoch_end
         epoch_end_logged_outputs = self.trainer._logger_connector.update_eval_epoch_metrics()
+        logger_metrics = dict(ChainMap(*logged_outputs))
+        logger_metrics.update(epoch_end_logged_outputs)
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:185]:
```suggestion
```

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 4]
```diff
@@ -765,13 +766,27 @@ def test_epoch_end(self, outputs) -> None:
...
+        "foo/dataloader_idx_0",
+        "foo/dataloader_idx_1",
+        "foobar",
+    }
+    assert set(mock_log_metrics.mock_calls[0].kwargs["metrics"]) == {
+        "foo/dataloader_idx_0",
+        "foo/dataloader_idx_1",
+        "foobar",
+        "epoch",
+    }
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:789]:
```suggestion
    cb_metrics = set(trainer.callback_metrics)
    assert cb_metrics == {"foo/dataloader_idx_0", "foo/dataloader_idx_1", "foobar"}
    assert set(mock_log_metrics.mock_calls[0].kwargs["metrics"]) == cb_metrics + {"epoch"}
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:190]: it will, but since epoch end metrics are duplicated here, I think the current one might be a little better performance-wise.

File: pytorch_lightning/utilities/imports.py [hunk 5]
```diff
@@ -89,6 +89,7 @@ def _compare_version(package: str, op: Callable, version: str, use_base_version:
 
 _IS_WINDOWS = platform.system() == "Windows"
 _IS_INTERACTIVE = hasattr(sys, "ps1")  # https://stackoverflow.com/a/64523765
+_PYTHON_GRATER_EQUAL_3_8_0 = _compare_version("python", operator.ge, "3.8.0")
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:92]:
```suggestion
_PYTHON_GREATER_EQUAL_3_8_0 = _compare_version("python", operator.ge, "3.8.0")
```

Comment (rohitgr7) [reply pytorch_lightning/utilities/imports.py]: ah! thank you

File: tests/loops/test_evaluation_loop.py [hunk 6]
```diff
@@ -48,7 +48,7 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
 def test_log_epoch_metrics_before_on_evaluation_end(update_eval_epoch_metrics_mock, tmpdir):
     """Test that the epoch metrics are logged before the `on_evaluation_end` hook is fired."""
     order = []
-    update_eval_epoch_metrics_mock.side_effect = lambda: order.append("log_epoch_metrics")
+    update_eval_epoch_metrics_mock.side_effect = lambda x: order.append("log_epoch_metrics")
```
Comment (awaelchli) [inline tests/loops/test_evaluation_loop.py:51]:
```suggestion
    update_eval_epoch_metrics_mock.side_effect = lambda _: order.append("log_epoch_metrics")
```

**Review by awaelchli — APPROVED** [review]
low confidence review

**Review by krshrimali — APPROVED** [review]
LGTM, thanks!
